File: sdk/python/agentpayments_python/grant_store.py
"""
Grant stores for durable paid-key persistence (P0 #5).

Once a key is added to a grant store it is never re-scanned on-chain, making
paid access durable even after the vendor wallet accumulates 100+ newer
transactions that would push the original payment out of the scan window.

Usage — pass to any adapter as ``grant_store``:

    from agentpayments_python.grant_store import FileGrantStore

    # Django settings.py
    AGENTPAYMENTS_GRANT_STORE = FileGrantStore("/var/data/agp_grants.json")

    # FastAPI / Flask constructor arg
    register_agentpayments(app, ..., grant_store=FileGrantStore("/var/data/agp_grants.json"))

Grant store interface (implement your own for Redis, Postgres, etc.):

    class GrantStore(Protocol):
        def has(self, agent_key: str) -> bool: ...
        def add(self, agent_key: str, expires_at: float | None = None, tier: str | None = None) -> None: ...
        def revoke(self, agent_key: str) -> None: ...

``add``'s ``expires_at``/``tier`` arguments are optional — a custom store
whose ``add`` only accepts ``agent_key`` will break if a caller passes them
as keywords; this is the one Python-side interface change to be aware of
when upgrading a custom store (see CHANGELOG). ``revoke`` is called by the
vendor's own code (e.g. an admin route) to cut off a specific key early; the
adapters never call it — ``has()`` on the built-in stores already returns
False for a revoked or expired grant.
"""

from __future__ import annotations

import json
import os
import threading
import time
from pathlib import Path


def _is_expired(grant: dict) -> bool:
    expires_at = grant.get("expires_at")
    return expires_at is not None and time.time() > expires_at


class MemoryGrantStore:
    """In-memory grant store. Does not survive restarts."""

    def __init__(self) -> None:
        self._grants: dict[str, dict] = {}
        self._lock = threading.Lock()

    def has(self, agent_key: str) -> bool:
        with self._lock:
            grant = self._grants.get(agent_key)
            if not grant or grant.get("revoked") or _is_expired(grant):
                return False
            return True

    def add(self, agent_key: str, expires_at: float | None = None, tier: str | None = None) -> None:
        with self._lock:
            self._grants[agent_key] = {"expires_at": expires_at, "tier": tier, "revoked": False}

    def revoke(self, agent_key: str) -> None:
        with self._lock:
            existing = self._grants.get(agent_key, {"expires_at": None, "tier": None})
            self._grants[agent_key] = {**existing, "revoked": True}


class FileGrantStore:
    """
    File-backed grant store. Persists grants to a JSON file so they survive
    process restarts. Writes are atomic (write to temp file, os.replace).

    Reads the legacy on-disk format (a plain JSON array of key strings) as
    pre-existing permanent grants, so upgrading does not invalidate a
    vendor's existing grants file. Writes the newer object-map format going
    forward.

    Not suitable for multi-process deployments — use a database or Redis there.
    """

    def __init__(self, path: str | Path) -> None:
        self._path = Path(path).resolve()
        self._grants: dict[str, dict] = {}
        self._lock = threading.Lock()
        self._load()

    def _load(self) -> None:
        try:
            parsed = json.loads(self._path.read_text())
            if isinstance(parsed, list):
                # Legacy format: array of key strings, each a permanent grant.
                for key in parsed:
                    self._grants[key] = {"expires_at": None, "tier": None, "revoked": False}
            elif isinstance(parsed, dict):
                for key, grant in parsed.items():
                    self._grants[key] = {
                        "expires_at": grant.get("expires_at"),
                        "tier": grant.get("tier"),
                        "revoked": bool(grant.get("revoked")),
                    }
        except FileNotFoundError:
            pass  # will be created on first write

    def _save(self) -> None:
        self._path.parent.mkdir(parents=True, exist_ok=True)
        tmp = self._path.with_suffix(".tmp")
        tmp.write_text(json.dumps(self._grants, indent=2, sort_keys=True))
        os.replace(tmp, self._path)

    def has(self, agent_key: str) -> bool:
        with self._lock:
            grant = self._grants.get(agent_key)
            if not grant or grant.get("revoked") or _is_expired(grant):
                return False
            return True

    def add(self, agent_key: str, expires_at: float | None = None, tier: str | None = None) -> None:
        with self._lock:
            self._grants[agent_key] = {"expires_at": expires_at, "tier": tier, "revoked": False}
            self._save()

    def revoke(self, agent_key: str) -> None:
        with self._lock:
            existing = self._grants.get(agent_key, {"expires_at": None, "tier": None})
            self._grants[agent_key] = {**existing, "revoked": True}
            self._save()

File: sdk/node/grant-store.js
/**
 * Grant stores for durable paid-key persistence (P0 #5).
 *
 * The 10-minute payment verification cache is NOT persistence — once a wallet
 * receives 100+ newer transactions the original payment can no longer be found
 * in the last-100-signatures scan. A grant store fixes this: once a key is
 * verified it is recorded permanently (or until it expires/is revoked) and
 * never needs re-scanning while the grant is valid.
 *
 * Usage — pass to agentPaymentsGate:
 *
 *   const { agentPaymentsGate } = require('@agentpayments/node');
 *   const { FileGrantStore } = require('@agentpayments/node/grant-store');
 *
 *   app.use(agentPaymentsGate({
 *     ...config,
 *     grantStore: new FileGrantStore('./data/grants.json'),
 *   }));
 *
 * Grant store interface (implement your own for Redis, Postgres, etc.):
 *
 *   interface GrantStore {
 *     has(agentKey: string): boolean | Promise<boolean>
 *     add(agentKey: string, grant?: { expiresAt?: number|null, tier?: string|null }): void | Promise<void>
 *     revoke(agentKey: string): void | Promise<void>
 *   }
 *
 * `add`'s second argument is optional — a store that only implements
 * `add(agentKey)` keeps working exactly as before (extra args are ignored),
 * granting indefinite access. `revoke` is called by the vendor's own code
 * (e.g. an admin route) to cut off a specific key early; the gate itself
 * never calls it — `has()` on the built-in stores already returns false for
 * a revoked or expired grant.
 */

'use strict';

const fs = require('node:fs');
const path = require('node:path');

function isExpired(grant) {
  return grant.expiresAt != null && Date.now() > grant.expiresAt;
}

/**
 * In-memory grant store. Survives restarts only if combined with a persistent
 * backing store. Useful as a default / for testing.
 */
class MemoryGrantStore {
  constructor() {
    this._grants = new Map();
  }
  has(key) {
    const grant = this._grants.get(key);
    if (!grant || grant.revoked || isExpired(grant)) return false;
    return true;
  }
  add(key, grant = {}) {
    this._grants.set(key, { expiresAt: grant.expiresAt ?? null, tier: grant.tier ?? null, revoked: false });
  }
  revoke(key) {
    const existing = this._grants.get(key);
    this._grants.set(key, { ...(existing || { expiresAt: null, tier: null }), revoked: true });
  }
}

/**
 * File-backed grant store. Persists grants to a JSON file on disk so they
 * survive server restarts. Write is atomic (temp-file + rename).
 *
 * Reads the legacy on-disk format (a plain JSON array of key strings) as
 * pre-existing permanent grants, so upgrading does not invalidate a vendor's
 * existing grants file. Writes the newer object-map format going forward.
 *
 * Not suitable for multi-process deployments — use Redis or a database there.
 */
class FileGrantStore {
  /**
   * @param {string} filePath  Absolute or relative path to the JSON grants file.
   *                           The file is created if it does not exist.
   */
  constructor(filePath) {
    this._path = path.resolve(filePath);
    this._grants = new Map();
    this._load();
  }

  _load() {
    try {
      const raw = fs.readFileSync(this._path, 'utf8');
      const parsed = JSON.parse(raw);
      if (Array.isArray(parsed)) {
        // Legacy format: array of key strings, each a permanent grant.
        parsed.forEach((k) => this._grants.set(k, { expiresAt: null, tier: null, revoked: false }));
      } else if (parsed && typeof parsed === 'object') {
        for (const [key, grant] of Object.entries(parsed)) {
          this._grants.set(key, { expiresAt: grant.expiresAt ?? null, tier: grant.tier ?? null, revoked: Boolean(grant.revoked) });
        }
      }
    } catch (err) {
      if (err.code !== 'ENOENT') throw err; // only ignore missing file
    }
  }

  _save() {
    fs.mkdirSync(path.dirname(this._path), { recursive: true });
    const tmp = this._path + '.tmp';
    const obj = Object.fromEntries(this._grants);
    fs.writeFileSync(tmp, JSON.stringify(obj, null, 2), 'utf8');
    fs.renameSync(tmp, this._path);
  }

  has(key) {
    const grant = this._grants.get(key);
    if (!grant || grant.revoked || isExpired(grant)) return false;
    return true;
  }

  add(key, grant = {}) {
    this._grants.set(key, { expiresAt: grant.expiresAt ?? null, tier: grant.tier ?? null, revoked: false });
    this._save();
  }

  revoke(key) {
    const existing = this._grants.get(key);
    this._grants.set(key, { ...(existing || { expiresAt: null, tier: null }), revoked: true });
    this._save();
  }
}

module.exports = { MemoryGrantStore, FileGrantStore };
